### ISSUE TITLE
train_spec.max_steps is ignored by tf.estimator.train_and_evaluate()

### ISSUE THREAD

**Issue by tarokiritani** [tensorflow/tensorflow#21748]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.10.0-0-g656e7a2b34 1.10.0
- **Python version**: 3.6.4
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:  CUDA Version 9.1.85
- **GPU model and memory**: Tesla K80
- **Exact command to reproduce**: The following script

```
import tensorflow as tf
tf.logging.set_verbosity(20)
import math
import numpy as np

x = np.array([[1,2,3.3, 2,10], [2,4.6,45,5,3],[5,5,56,2,1],[1,2,3,4,100],[10,2,1,4,50]]).astype(np.float32)
y=np.array([0,0,0,1,1]).astype(np.float32)
batch_size = 2

train_input_fn = tf.estimator.inputs.numpy_input_fn(x={'x': x},
        y=y,
        batch_size=batch_size,
        num_epochs=1,
        shuffle=True
        )

eval_input_fn = tf.estimator.inputs.numpy_input_fn(x={'x': np.array([[1,2,3,1,4], [1,23,4,1,90]]).astype(np.float32)},
        y=np.array([0,1]).astype(np.float32),
        batch_size=batch_size,
        num_epochs=1,
        shuffle=False
        )

dnn = tf.estimator.DNNClassifier([32],
    feature_columns=[tf.feature_column.numeric_column('x', shape=[5])])

steps_per_epoch = math.ceil(len(x) / batch_size)
epochs = 3

train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn, max_steps=epochs * steps_per_epoch)
eval_spec = tf.estimator.EvalSpec(input_fn=eval_input_fn)

tf.estimator.train_and_evaluate(dnn, train_spec, eval_spec)
```

### Describe the problem
tf.estimator.train_and_evaluate() stops after 1 iteration with tf 1.10. It works fine with tf 1.9. Training should occur 9 times (=3 epochs x 3 steps per epoch) in total.

### Source code / logs

The traceback with tf v1.9.0-0-g25c197e023 1.9.0:
```
INFO:tensorflow:Using default config.
WARNING:tensorflow:Using temporary folder as model directory: /tmp/tmpgn5y57s6
INFO:tensorflow:Using config: {'_model_dir': '/tmp/tmpgn5y57s6', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_service': None, '_cluster_spec': <tensorflow.python.tr
aining.server_lib.ClusterSpec object at 0x7f29437f6908>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 
0, '_num_worker_replicas': 1}
INFO:tensorflow:Running training and evaluation locally (non-distributed).
INFO:tensorflow:Start train and evaluate loop. The evaluate will happen after 600 secs (eval_spec.throttle_secs) or training is finished.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
2018-08-21 01:06:30.556250: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 0 into /tmp/tmpgn5y57s6/model.ckpt.
INFO:tensorflow:loss = 11.228095, step = 1
INFO:tensorflow:Saving checkpoints for 3 into /tmp/tmpgn5y57s6/model.ckpt.
INFO:tensorflow:Loss for final step: 0.016822023.
INFO:tensorflow:Calling model_fn.
WARNING:tensorflow:Trapezoidal rule is known to produce incorrect PR-AUCs; please switch to "careful_interpolation" instead.
WARNING:tensorflow:Trapezoidal rule is known to produce incorrect PR-AUCs; please switch to "careful_interpolation" instead.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-08-21-01:06:31
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /tmp/tmpgn5y57s6/model.ckpt-3
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2018-08-21-01:06:32
INFO:tensorflow:Saving dict for global step 3: accuracy = 1.0, accuracy_baseline = 0.5, auc = 0.99999905, auc_precision_recall = 0.999999, average_loss = 0.27123037, global_step = 3, label/mean = 0.5, los
s = 0.54246074, precision = 1.0, prediction/mean = 0.4936065, recall = 1.0
...
---Information of steps 1-9 is shown.---
```

With tf 1.10, the program ends after step 3:
```
INFO:tensorflow:Using default config.
WARNING:tensorflow:Using temporary folder as model directory: /tmp/tmp686id5oc
INFO:tensorflow:Using config: {'_model_dir': '/tmp/tmp686id5oc', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f10547942b0>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}
INFO:tensorflow:Running training and evaluation locally (non-distributed).
INFO:tensorflow:Start train and evaluate loop. The evaluate will happen after every checkpoint. Checkpoint frequency is determined based on RunConfig arguments: save_checkpoints_steps None or save_checkpoints_secs 600.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
2018-08-21 01:17:17.979737: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 0 into /tmp/tmp686id5oc/model.ckpt.
INFO:tensorflow:loss = 6.8426137, step = 1
INFO:tensorflow:Saving checkpoints for 3 into /tmp/tmp686id5oc/model.ckpt.
INFO:tensorflow:Calling model_fn.
WARNING:tensorflow:Trapezoidal rule is known to produce incorrect PR-AUCs; please switch to "careful_interpolation" instead.
WARNING:tensorflow:Trapezoidal rule is known to produce incorrect PR-AUCs; please switch to "careful_interpolation" instead.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-08-21-01:17:19
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /tmp/tmp686id5oc/model.ckpt-3
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2018-08-21-01:17:19
INFO:tensorflow:Saving dict for global step 3: accuracy = 1.0, accuracy_baseline = 0.5, auc = 0.99999905, auc_precision_recall = 0.999999, average_loss = 0.371122, global_step = 3, label/mean = 0.5, loss = 0.742244, precision = 1.0, prediction/mean = 0.42290705, recall = 1.0
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 3: /tmp/tmp686id5oc/model.ckpt-3
INFO:tensorflow:Loss for final step: 1.7909913.
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tarokiritani** [comment]
In the code above, 'num_epochs' in tf.estimator.inputs.numpy_input_fn() actually determines the total number of steps, although the [documentation](https://www.tensorflow.org/versions/r1.10/api_docs/python/tf/estimator/train_and_evaluate) says "the only supported stop condition for model training is train_spec.max_steps."

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tarokiritani** [comment]
@danielhomola It seems the number of steps is determined by the input function. Change the `num_epochs` in tf.estimator.inputs.numpy_input_fn(). If you use tf.data.Dataset, .repeat() will change the epoch number.

**Comment by danielhomola** [comment]
I found this to be not the case.. I increased repeat in my script and train_eval still terminated after one epoch..

**Comment by tarokiritani** [comment]
@danielhomola Does my snippet above behave differently on your system? `num_epochs` of the `train_input_fn` changes nothing?

**Comment by danielhomola** [comment]
I used a custom input_fn of my own not `train_input_fn`..

**Comment by KuenstlicheIntelligenz** [comment]
Did anyone solved it? Having the same problem...

**Comment by yangyanli** [comment]
Having the same problem here. If I set steps in EvalSpec, then train will continue after steps of eval is done. I didn't check the code, but it seems that the end-of-input exception of *eval* stops *train*...

**Comment by KuenstlicheIntelligenz** [comment]
Ive digged into the code a few days ago and somewhere was a TODO that they have to fix the OutOfRangeError after fixing some other bug.

For now i restart the training after each epoch and reload the whole model again which takes quite some time. Maybe its worth to change the version of tf...

**Comment by tarokiritani** [comment]
@KuenstlicheIntelligenz Thanks for checking the source. In tf 1.12, I now use tf.data.Dataset, and the epoch number is determined by the dataset, not by the train_spec.

**Comment by ShakaHua** [comment]
I used tf 1.8.0 and python2. It seems that max_steps determines the total steps, not by the epoch numbers in TrainSpec.

**Comment by tilakrayal** [comment]
@tarokiritani ,
We see that you are using tf version 1.4.0, 1.x is not actively supported, please update to 2.x and let us know if you are facing same issue.Thanks!
